### PR TITLE
ADEN-4833 Add src=rec for PageFair ads

### PR DESCRIPTION
**Pull request by dianafa**
https://wikia-inc.atlassian.net/browse/ADEN-4833

### CODE REVIEW

**Review by mleczey — APPROVED** [review]

**Comment by nandy-andy** [comment]
The code changes are readable and I can read from them that it's going to do what we want it to do. It's covered with unit tests :+1: ...but I wonder why we did it differently than SourcePoint. I'd like to take a 2nd look tomorrow. Maybe we'll pair-CR, @dianafa? :-) Oh, and Jenkins failed :-(

**Comment by dianafa** [comment]
Jenkins fails not related to this PR.

Generally the reason was that I didn't like the way we structured SourcePoint recovery code. Before it was a general "recovery", not prepared to handle more than one recovery provider.
I didn't want to extend `recoverySlots` array or create a new one, for PF recovery in provider/directGpt.js (https://github.com/Wikia/app/pull/12567/files#diff-17d2543e01f9dbebdd79cc24d8a5674eR55) that is general gpt file, with configuration specific for pageFair. I'm not sure if it's clear what I mean though, let's pair-cr on that :)

File: extensions/wikia/ARecoveryEngine/js/PageFair/pageFair.js [hunk 1]
```diff
@@ -19,19 +20,27 @@ define('ext.wikia.aRecoveryEngine.recovery.pageFair', [
...
+		var result = isPageFairRecoveryEnabled() && recoverableSlots.indexOf(slotName) !== -1;
 
 		log(['isSlotRecoverable', result], log.levels.info, logGroup);
 		return result;
 	}
 
+	function isBlocking() {
+		var isBlocking = !!(win.ads && win.ads.runtime.pf && win.ads.runtime.pf.blocking);
+
+		log(['isBlocking', isBlocking], 'debug', logGroup);
```
Comment (nandy-andy) [inline extensions/wikia/ARecoveryEngine/js/PageFair/pageFair.js]: We can also replace `'debug'` in all places with `log.levels.debug` as :cherries: ;-)

File: extensions/wikia/ARecoveryEngine/js/PageFair/pageFair.js [hunk 2]
```diff
@@ -19,19 +20,27 @@ define('ext.wikia.aRecoveryEngine.recovery.pageFair', [
 	}
 
 	function isSlotRecoverable(slotName) {
-		var result = isPageFairRecoveryEnabled() && !!recoverableSlots[slotName];
+		var result = isPageFairRecoveryEnabled() && recoverableSlots.indexOf(slotName) !== -1;
```
Comment (nandy-andy) [inline extensions/wikia/ARecoveryEngine/js/PageFair/pageFair.js]: What do you think of renaming `isPageFairRecoveryEnabled()` to just `isEnabled()`?

File: extensions/wikia/ARecoveryEngine/js/sourcePoint.js [hunk 3]
```diff
@@ -51,12 +60,17 @@ define('ext.wikia.aRecoveryEngine.recovery.sourcePointHelper', [
...
+		var isBlocking = !!(win.ads && win.ads.runtime.sp && win.ads.runtime.sp.blocking);
+
+		log(['isBlocking', isBlocking], 'debug', logGroup);
+		return isBlocking;
 	}
 
-	function isSourcePointRecoverable(slotName, recoverableSlots) {
-		return isSourcePointRecoveryEnabled() && recoverableSlots.indexOf(slotName) !== -1;
+	function isSlotRecoverable(slotName) {
+		var result = isSourcePointRecoveryEnabled() && recoverableSlots.indexOf(slotName) !== -1;
```
Comment (nandy-andy) [inline extensions/wikia/ARecoveryEngine/js/sourcePoint.js]: What do you think of renaming `isSourcePointRecoveryEnabled()` to just `isEnabled()`? :-p

File: extensions/wikia/ARecoveryEngine/templates/ARecoveryEngineApiController_getBootstrap.mustache [hunk 4]
```diff
@@ -19,7 +19,7 @@
 					document.getElementById('WikiaArticle').style.display = 'none';
 					document.getElementById('WikiaArticleMsg').style.display = 'block';
 					window.addEventListener('load', function () {
-						require(['ext.wikia.aRecoveryEngine.recovery.sourcePointHelper'], function (helper) {
+						require(['ext.wikia.aRecoveryEngine.recovery.sourcePoint'], function (helper) {
```
Comment (nandy-andy) [inline extensions/wikia/ARecoveryEngine/templates/ARecoveryEngineApiController_getBootstrap.mustache:22]: We'll most likely need a fallback until cache re-validates.

File: extensions/wikia/ARecoveryEngine/templates/ARecoveryEngineApiController_getBootstrap.mustache [hunk 5]
```diff
@@ -37,7 +37,7 @@
 			window._sp_.bootstrap('{{domain}}/api/v1/ARecoveryEngine/delivery');
 
 			window.addEventListener('load', function () {
-				require(['ext.wikia.aRecoveryEngine.recovery.sourcePointHelper'], function (helper) {
+				require(['ext.wikia.aRecoveryEngine.recovery.sourcePoint'], function (helper) {
```
Comment (nandy-andy) [inline extensions/wikia/ARecoveryEngine/templates/ARecoveryEngineApiController_getBootstrap.mustache:40]: What will happen if a page will be cached and ask for module which doesn't exist anymore?

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js [hunk 6]
```diff
@@ -52,7 +52,7 @@ define('ext.wikia.adEngine.lookup.lookupFactory', [
 			module.call(context.targeting.skin || 'mercury', onResponse);
 			called = true;
 
-			helper.addOnBlockingCallback(onResponseCallbacks.start);
+			recoveryHelper.addOnBlockingCallback(onResponseCallbacks.start);
```
Comment (nandy-andy) [inline extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: How does PF cooperates with RTBs?

File: extensions/wikia/AdEngine/js/SlotTweaker.js [hunk 7]
```diff
@@ -3,11 +3,10 @@ define('ext.wikia.adEngine.slotTweaker', [
 	'ext.wikia.adEngine.domElementTweaker',
 	'ext.wikia.adEngine.messageListener',
 	'ext.wikia.adEngine.slot.adSlot',
-	'ext.wikia.aRecoveryEngine.recovery.sourcePointHelper',
 	'wikia.document',
 	'wikia.log',
 	'wikia.window'
-], function (DOMElementTweaker, messageListener, adSlot, recoveryHelper, doc, log, win) {
+], function (DOMElementTweaker, messageListener, adSlot, doc, log, win) {
```
Comment (nandy-andy) [inline extensions/wikia/AdEngine/js/SlotTweaker.js:9]: Unused dependency, good catch :+1:

File: extensions/wikia/AdEngine/js/context/adContext.js [hunk 8]
```diff
@@ -111,9 +111,10 @@ define('ext.wikia.adEngine.adContext', [
 		// SourcePoint detection integration
 		if (!noExternals && context.opts.sourcePointDetectionUrl) {
 			context.opts.sourcePointDetection = (context.targeting.skin === 'oasis' &&
-			geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionCountries));
+				geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionCountries));
+
 			context.opts.sourcePointDetectionMobile = (context.targeting.skin === 'mercury' &&
-			geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionMobileCountries));
+				geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionMobileCountries));
```
Comment (nandy-andy) [inline extensions/wikia/AdEngine/js/context/adContext.js:114]: Is it an unnecessary tab?

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js [hunk 9]
```diff
@@ -2,10 +2,10 @@
 define('ext.wikia.adEngine.lookup.lookupFactory', [
 	'ext.wikia.adEngine.adContext',
 	'ext.wikia.adEngine.adTracker',
-	'ext.wikia.aRecoveryEngine.recovery.sourcePointHelper',
+	'ext.wikia.aRecoveryEngine.recovery.sourcePoint',
 	'wikia.lazyqueue',
 	'wikia.log'
-], function (adContext, adTracker, helper, lazyQueue, log) {
+], function (adContext, adTracker, recoveryHelper, lazyQueue, log) {
```
Comment (nandy-andy) [inline extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: What do you think of changing `recoveryHelper` in the functions arguments list to `sourcePoint` or `sourcePointRecovery`?

File: extensions/wikia/AdEngine/js/provider/gpt/googleTag.js [hunk 10]
```diff
@@ -62,10 +62,11 @@ define('ext.wikia.adEngine.provider.gpt.googleTag', [
 
 		var gads = doc.createElement('script'),
 			node = doc.getElementsByTagName('script')[0],
-			// load GPT when API not ready and recovery is non blocking or recovery is pageFair (never blocking)
-			pageFairRecovery = pageFair && pageFair.isPageFairRecoveryEnabled(),
+			pageFairRecoveryisBlocking = pageFair && pageFair.isBlocking(),
```
Comment (nandy-andy) [inline extensions/wikia/AdEngine/js/provider/gpt/googleTag.js]: How about `pageFairRecoveryisBlocking` -> `pageFairRecoveryIsBlocking` or `isPageFairRecoveryBlocking`?

File: extensions/wikia/AdEngine/js/context/adContext.js (hunk 8 above)
Comment (dianafa) [reply extensions/wikia/AdEngine/js/context/adContext.js:114]: imo is necessary

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js (hunk 9 above)
Comment (dianafa) [reply extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: yep, was thinking about that

File: extensions/wikia/AdEngine/js/provider/gpt/googleTag.js (hunk 10 above)
Comment (dianafa) [reply extensions/wikia/AdEngine/js/provider/gpt/googleTag.js]: better: `isPageFairBlocking` - it's not a recovery that blocks. PF just detects it. Generally I don't like any of those names...

File: extensions/wikia/ARecoveryEngine/js/PageFair/pageFair.js (hunk 1 above)
Comment (dianafa) [reply extensions/wikia/ARecoveryEngine/js/PageFair/pageFair.js]: thanks! :)

**Review by nandy-andy — APPROVED** [review]

File: extensions/wikia/ARecoveryEngine/js/sourcePoint.js [hunk 11]
```diff
@@ -43,20 +52,25 @@ define('ext.wikia.aRecoveryEngine.recovery.sourcePointHelper', [
...
 
-		log(['isSourcePointRecoveryEnabled', enabled, 'debug', logGroup]);
+		log(['isEnabled', enabled, log.levels.debug, logGroup]);
 		return enabled;
 	}
 
 	function isBlocking() {
-		log(['isBlocking', !!(win.ads && win.ads.runtime.sp && win.ads.runtime.sp.blocking)], 'debug', logGroup);
-		return isSourcePointRecoveryEnabled() && !!(win.ads && win.ads.runtime.sp && win.ads.runtime.sp.blocking);
+		var isBlocking = !!(win.ads && win.ads.runtime.sp && win.ads.runtime.sp.blocking);
```
Comment (mleczey) [inline extensions/wikia/ARecoveryEngine/js/sourcePoint.js:63]: Just checking, I assume we don't need `isEnabled` anymore here?

File: extensions/wikia/AdEngine/js/lookup/prebid/prebid.js [hunk 12]
```diff
@@ -21,13 +21,6 @@ define('ext.wikia.adEngine.lookup.prebid', [
 	win
 ) {
 	'use strict';
-
-	/*
-	 * When updating prebid.js (https://github.com/prebid/Prebid.js/) to a new version
-	 * remember about the additional [320, 480] slot size, see:
-	 * https://github.com/Wikia/app/pull/12269/files#diff-5bbaaa809332f9adaddae42c8847ae5bR6015
-	 */
-
```
Comment (mleczey) [inline extensions/wikia/AdEngine/js/lookup/prebid/prebid.js:30]: I think this change is not related. Can we get rid of it?

File: extensions/wikia/AdEngine/js/provider/gpt/helper.js [hunk 13]
```diff
@@ -91,7 +99,7 @@ define('ext.wikia.adEngine.provider.gpt.helper', [
 				}
 			}
 
-			if (shouldPushRecoverableAd) {
+			if (isRecoveryEnabled && isBlocking && adIsRecoverable) {
```
Comment (mleczey) [inline extensions/wikia/AdEngine/js/provider/gpt/helper.js]: Would it be possible to extract method here, or simplify conditions logic in changed lines in this file?

File: extensions/wikia/AdEngine/js/lookup/prebid/prebid.js (hunk 12 above)
Comment (dianafa) [reply extensions/wikia/AdEngine/js/lookup/prebid/prebid.js:30]: I just encountered this old piece of code and I want to clean this up :)

File: extensions/wikia/ARecoveryEngine/js/sourcePoint.js (hunk 11 above)
Comment (dianafa) [reply extensions/wikia/ARecoveryEngine/js/sourcePoint.js:63]: yes, we sometimes want to check if user is blocking even though recovery is enabled

**Comment by mleczey** [comment]
Sorry, I just copied wrong link to pull request. Reference added by me has nothing to do with this.

**Review by mleczey — APPROVED** [review]

**Review by mleczey — APPROVED** [review]
